ouroboros-consensus-protocol/src/unstable-protocol-testlib/Test/Ouroboros/Consensus/Protocol/Praos/Header.hs

```diff
@@ -0,0 +1,456 @@
+    hbPrev <- BlockHash . HashHeader <$> genHash
+    let hbVk = VKey $ deriveVerKeyDSIGN coldSignKey
+    hbBodySize <- choose (1000, 90000)
+    hbBodyHash <- genHash
+    (hbOCert, kesPeriod) <- genCert hbSlotNo context
+    let hbProtVer = protocolVersionZero
+        headerBody = HeaderBody{..}
+    pure $ (headerBody, kesPeriod)
+  where
+    GeneratorContext{coldSignKey} = context
+
+genLeadingSlot :: GeneratorContext -> Gen (SlotNo, VRF.CertifiedVRF VRF.PraosVRF InputVRF, VRF.VerKeyVRF VRF.PraosVRF)
+genLeadingSlot context = do
+    slotNo <- SlotNo . getPositive <$> arbitrary `suchThat` isLeader
+    let rho' = mkInputVRF slotNo nonce
+        hbVrfRes = VRF.evalCertified () rho' vrfSignKey
+        hbVrfVk = deriveVerKeyVRF vrfSignKey
+    pure (slotNo, hbVrfRes, hbVrfVk)
+  where
+    isLeader n =
+        let slotNo = SlotNo . getPositive $ n
+            rho' = mkInputVRF slotNo nonce
+            certified = VRF.evalCertified () rho' vrfSignKey
+         in checkLeaderNatValue (vrfLeaderValue (Proxy @StandardCrypto) certified) sigma activeSlotCoeff
+    sigma = 1
+    GeneratorContext{vrfSignKey, nonce, activeSlotCoeff} = context
+
+protocolVersionZero :: ProtVer
+protocolVersionZero = ProtVer versionZero 0
+  where
+    versionZero :: Version
+    versionZero = natVersion @0
+
+genCert :: SlotNo -> GeneratorContext -> Gen (OCert StandardCrypto, KESPeriod)
+genCert slotNo context = do
+    let ocertVkHot = KES.deriveVerKeyKES kesSignKey
+        poolId = coerce $ hashKey $ VKey $ deriveVerKeyDSIGN coldSignKey
+        ocertN = fromMaybe 0 $ Map.lookup poolId ocertCounters
+    ocertKESPeriod <- genValidKESPeriod slotNo praosSlotsPerKESPeriod
+    let ocertSigma = signedDSIGN @StandardCrypto coldSignKey (OCertSignable ocertVkHot ocertN ocertKESPeriod)
+    pure (OCert{..}, ocertKESPeriod)
+  where
+    GeneratorContext{kesSignKey, praosSlotsPerKESPeriod, coldSignKey, ocertCounters} = context
+
+genValidKESPeriod :: SlotNo -> Word64 -> Gen KESPeriod
+genValidKESPeriod slotNo praosSlotsPerKESPeriod =
+    pure $ KESPeriod $ fromIntegral $ unSlotNo slotNo `div` praosSlotsPerKESPeriod
+
+genKESPeriodAfterLimit :: SlotNo -> Word64 -> Gen KESPeriod
+genKESPeriodAfterLimit slotNo praosSlotsPerKESPeriod =
+    KESPeriod . fromIntegral <$> arbitrary `suchThat` (> currentKESPeriod)
+  where
+    currentKESPeriod = unSlotNo slotNo `div` praosSlotsPerKESPeriod
+
+genSlotAfterKESPeriod :: Word64 -> Word64 -> Word64 -> Gen SlotNo
+genSlotAfterKESPeriod ocertKESPeriod praosMaxKESEvo praosSlotsPerKESPeriod = do
+    -- kp_ < c0_ +  praosMaxKESEvo
+    -- ! =>
+    -- kp >=  c0_ +  praosMaxKESEvo
+    -- c0 <=  kp -  praosMaxKESEvo
+    SlotNo <$> arbitrary `suchThat` (> threshold)
+  where
+    threshold = (ocertKESPeriod + praosMaxKESEvo + 1) * praosSlotsPerKESPeriod
+
+genHash :: Gen (Hash Blake2b_256 a)
+genHash = coerce . hashWith id <$> gen32Bytes
+
+gen32Bytes :: Gen ByteString
+gen32Bytes = BS.pack <$> vectorOf 32 arbitrary
```

Comment:
Same as above, maybe even using some newtypes.

Reply:
If you don't mind I would rather avoid that dance: these generators are used only locally and adding a `newtype` declaration at this stage seems overkill to me.